Mathlib/NumberTheory/ModularForms/ArithmeticSubgroups.lean

```diff
@@ -51,3 +55,24 @@ instance IsArithmetic.finiteIndex_comap (𝒢 : Subgroup (GL (Fin 2) ℝ)) [IsAr
   ⟨𝒢.index_comap (mapGL (R := ℤ) ℝ) ▸ IsArithmetic.is_commensurable.1⟩
 
 end Subgroup
+
+namespace Matrix.SpecialLinearGroup
+
+lemma isInducing_toGL : Topology.IsInducing (mapGL ℝ : SL(2, ℤ) → GL (Fin 2) ℝ) := by
+  refine .of_comp continuous_of_discreteTopology Units.continuous_val ?_
+  refine (Topology.IsInducing.matrix_map ?_).comp ⟨rfl⟩
+  exact Topology.IsEmbedding.toIsInducing (Isometry.isEmbedding fun _ _ ↦ rfl)
+
+lemma isEmbedding_toGL : Topology.IsEmbedding (mapGL ℝ : SL(2, ℤ) → GL (Fin 2) ℝ) :=
+  ⟨isInducing_toGL, mapGL_injective⟩
+
+instance discreteTopology_SL2ℤ : DiscreteTopology 𝒮ℒ :=
+  isEmbedding_toGL.toHomeomorph.discreteTopology
+
+lemma isClosed_SL2ℤ : IsClosed (𝒮ℒ : Set (GL (Fin 2) ℝ)) :=
+  Subgroup.isClosed_of_discrete
+
+lemma isClosedEmbedding_toGL : Topology.IsClosedEmbedding (mapGL ℝ : SL(2, ℤ) → GL (Fin 2) ℝ) :=
+  ⟨isEmbedding_toGL, isClosed_SL2ℤ⟩
```

Comment:
This whole section feels too specialized: you could at least allow an arbitrary index fintype. 

I also wanted to suggest that, in the first two statements, you could replace `(ℤ, ℝ)` by any pair `(A, B)` with `[NormedRing A] [NormedRing B] [Algebra A B] (h : IsInducing (algebraMap A B))`, but unfortunately the topology on `SL(n, A)` is incorrectly defined as the one inherited from `M_n(A)` instead of `GL(n, A)`, so I'm afraid that statement is not true for Mathlib's definitions.
Could you at least generalize to `[DiscreteTopology A]` (this way your proofs shouldn't break) and mention that it shouldn't be necessary (if you agree, it is also possible that I am missing something obvious).

Reply:
> Why "incorrectly"? For `SL(n)`, the entries of the inverse are polynomials in the entries of the original matrix (because the determinant is 1 you don't need to invert it), so inversion is continuous in the topology given by the embedding into M_n (which would not be true in GL_n of course). 

:facepalm: Sorry, you are right of course, I had a feeling of "this looks wrong" coming from the case of GL_n and didn't think more about it... 

Indeed I think the API you mention is good to have.